flax/core/lift.py

```diff
@@ -222,12 +222,32 @@ def swap(target):
   return transform((from_kind, to_kind), swap, swap, mutable=True)
 
 
+@dataclass(frozen=True)
+class In(Generic[T]):
+  axis: T 
+
+@dataclass(frozen=True)
+class Out(Generic[T]):
+  axis: T
+
+
+def _split_in_out_axes(xs: Mapping[KindFilter, Any]):
+  unpack = lambda v: v.axis if isinstance(v, (In, Out)) else v
+  in_axes = {k: unpack(v) for k, v in xs.items() if not isinstance(v, Out)}
+  out_axes = {k: unpack(v) for k, v in xs.items() if not isinstance(v, In)}
+  return in_axes, out_axes
+
+
+Axis = Optional[int]
+InOutAxis = Union[Axis, In[Axis], Out[Axis]]
+
+
 def vmap(fn: Callable[..., Any],
-         variable_in_axes: Mapping[KindFilter, Optional[int]],
-         variable_out_axes: Mapping[KindFilter, Optional[int]],
+         variable_axes: Mapping[KindFilter, InOutAxis],
          split_rngs: Mapping[KindFilter, bool],
          in_axes=0, out_axes=0, axis_size=None) -> Callable[..., Any]:
   """Wraps jax.vmap."""
```

Comment:
Probably not necessary for this PR but we should probably have pretty detailed docstrings for these lifted transformations.